Get rid of `-fsigned-char`

We currently set `-fsigned-char` for all DevilutionX code here:

https://github.com/diasurgical/devilutionX/blob/79b330825d11aefc62fc01f31d7b084193f8147c/CMake/functions/devilutionx_library.cmake#L47-L48

Our code should not rely on the signedness of char.

I think it already doesn't but let's test with `-funsigned-char` before removing the flag.

There are a couple of compilation errors with `-funsigned-char`:

```
/home/gleb/devilutionX/Source/monster.cpp: In function ‘void devilution::{anonymous}::SnakeAi(devilution::Monster&)’:
/home/gleb/devilutionX/Source/monster.cpp:2593:38: error: narrowing conversion of ‘-1’ from ‘int’ to ‘char’ [-Wnarrowing]
 2593 |         char pattern[6] = { 1, 1, 0, -1, -1, 0 };
      |                                      ^~
/home/gleb/devilutionX/Source/monster.cpp:2593:42: error: narrowing conversion of ‘-1’ from ‘int’ to ‘char’ [-Wnarrowing]
 2593 |         char pattern[6] = { 1, 1, 0, -1, -1, 0 };
      |                                          ^~
/home/gleb/devilutionX/Source/monster.cpp:2608:55: warning: comparison is always false due to limited range of data type [-Wtype-limits]
 2608 |                         if (pattern[monster.goalVar1] == -1)
      |                             ~~~~~~~~~~~~~~~~~~~~~~~~~~^~~~~
[48/232] Building CXX object Source/CMakeFiles/libdevilutionx.dir/stores.cpp.o
/home/gleb/devilutionX/Source/stores.cpp: In function ‘void devilution::{anonymous}::DrawSSlider(const devilution::Surface&, int, int)’:
/home/gleb/devilutionX/Source/stores.cpp:209:27: warning: comparison is always true due to limited range of data type [-Wtype-limits]
  209 |         if (stextscrlubtn != -1)
      |             ~~~~~~~~~~~~~~^~~~~
/home/gleb/devilutionX/Source/stores.cpp:213:27: warning: comparison is always true due to limited range of data type [-Wtype-limits]
  213 |         if (stextscrldbtn != -1)
      |             ~~~~~~~~~~~~~~^~~~~
/home/gleb/devilutionX/Source/stores.cpp: In function ‘void devilution::{anonymous}::StoreSellItem()’:
/home/gleb/devilutionX/Source/stores.cpp:1528:28: warning: comparison is always true due to limited range of data type [-Wtype-limits]
 1528 |         if (storehidx[idx] >= 0)
      |             ~~~~~~~~~~~~~~~^~~~
[56/232] Building CXX object Source/CMakeFiles/libdevilutionx.dir/options.cpp.o
ninja: build stopped: subcommand failed.
```

--- a/CMake/functions/devilutionx_library.cmake
+++ b/CMake/functions/devilutionx_library.cmake
@@ -43,9 +43,6 @@ function(add_devilutionx_library NAME)
 
     # Warnings for devilutionX
     target_compile_options(${NAME} PUBLIC -Wall -Wextra -Wno-unused-parameter)
-
-    # For ARM and other default unsigned char platforms
-    target_compile_options(${NAME} PUBLIC -fsigned-char)
   endif()
 
   if(NOT WIN32 AND NOT APPLE AND NOT ${CMAKE_SYSTEM_NAME} STREQUAL FreeBSD)

--- a/Source/levels/gendung.cpp
+++ b/Source/levels/gendung.cpp
@@ -35,7 +35,7 @@ _setlevels setlvlnum;
 dungeon_type setlvltype;
 Point ViewPosition;
 uint_fast8_t MicroTileLen;
-char TransVal;
+int8_t TransVal;
 bool TransList[256];
 uint16_t dPiece[MAXDUNX][MAXDUNY];
 MICROS DPieceMicros[MAXTILES];
@@ -47,7 +47,7 @@ int8_t dPlayer[MAXDUNX][MAXDUNY];
 int16_t dMonster[MAXDUNX][MAXDUNY];
 int8_t dCorpse[MAXDUNX][MAXDUNY];
 int8_t dObject[MAXDUNX][MAXDUNY];
-char dSpecial[MAXDUNX][MAXDUNY];
+int8_t dSpecial[MAXDUNX][MAXDUNY];
 int themeCount;
 THEME_LOC themeLoc[MAXTHEMES];
 

--- a/Source/levels/gendung.h
+++ b/Source/levels/gendung.h
@@ -183,7 +183,7 @@ extern dungeon_type setlvltype;
 /** Specifies the player viewpoint X,Y-coordinates of the map. */
 extern DVL_API_FOR_TEST Point ViewPosition;
 extern uint_fast8_t MicroTileLen;
-extern char TransVal;
+extern int8_t TransVal;
 /** Specifies the active transparency indices. */
 extern bool TransList[256];
 /** Contains the piece IDs of each tile on the map. */
@@ -219,7 +219,7 @@ extern DVL_API_FOR_TEST int8_t dObject[MAXDUNX][MAXDUNY];
  * (e.g. "levels/l1data/l1s"). Note, the special tileset of Tristram (i.e.
  * "levels/towndata/towns") contains trees rather than arches.
  */
-extern char dSpecial[MAXDUNX][MAXDUNY];
+extern int8_t dSpecial[MAXDUNX][MAXDUNY];
 extern int themeCount;
 extern THEME_LOC themeLoc[MAXTHEMES];
 

--- a/Source/monster.cpp
+++ b/Source/monster.cpp
@@ -2590,7 +2590,7 @@ void SnotSpilAi(Monster &monster)
 
 void SnakeAi(Monster &monster)
 {
-	char pattern[6] = { 1, 1, 0, -1, -1, 0 };
+	int8_t pattern[6] = { 1, 1, 0, -1, -1, 0 };
 	if (monster.mode != MonsterMode::Stand || monster.activeForTicks == 0)
 		return;
 	Direction md = GetDirection(monster.position.tile, monster.position.last);

--- a/Source/nthread.cpp
+++ b/Source/nthread.cpp
@@ -30,10 +30,10 @@ namespace {
 
 SdlMutex MemCrit;
 bool nthread_should_run;
-char sgbSyncCountdown;
+int8_t sgbSyncCountdown;
 uint32_t turn_upper_bit;
 bool sgbTicsOutOfSync;
-char sgbPacketCountdown;
+int8_t sgbPacketCountdown;
 bool sgbThreadIsRunning;
 SdlThread Thread;
 

--- a/Source/stores.cpp
+++ b/Source/stores.cpp
@@ -33,7 +33,7 @@ namespace devilution {
 talk_id stextflag;
 
 int storenumh;
-char storehidx[48];
+int8_t storehidx[48];
 Item storehold[48];
 
 Item smithitem[SMITH_ITEMS];
@@ -110,9 +110,9 @@ int stextdown;
 /** Previous scoll position */
 int stextup;
 /** Count down for the push state of the scroll up button */
-char stextscrlubtn;
+int8_t stextscrlubtn;
 /** Count down for the push state of the scroll down button */
-char stextscrldbtn;
+int8_t stextscrldbtn;
 
 /** Remember current store while displaying a dialog */
 talk_id stextshold;

--- a/Source/stores.h
+++ b/Source/stores.h
@@ -52,7 +52,7 @@ extern talk_id stextflag;
 /** Current index into storehidx/storehold */
 extern DVL_API_FOR_TEST int storenumh;
 /** Map of inventory items being presented in the store */
-extern char storehidx[48];
+extern int8_t storehidx[48];
 /** Copies of the players items as presented in the store */
 extern DVL_API_FOR_TEST Item storehold[48];
 

--- a/test/utf8_test.cpp
+++ b/test/utf8_test.cpp
@@ -74,9 +74,11 @@ TEST(Utf8CodeUnits, BasicLatin)
 
 	EXPECT_FALSE(IsBasicLatin('\x7F')) << "ASCII Control character DEL is not a Basic Latin symbol";
 
-	for (char x = '\x80'; x <= '\xFF'; x++) {
+	// Tests '\xFF' separately to avoid infinite loop on platforms with unsigned char.
+	for (char x = '\x80'; x < '\xFF'; x++) {
 		EXPECT_FALSE(IsBasicLatin(x)) << "Multibyte Utf8 code units are not Basic Latin symbols";
 	}
+	EXPECT_FALSE(IsBasicLatin('\xFF')) << "Multibyte Utf8 code units are not Basic Latin symbols";
 }
 
 } // namespace

